index/segment/fst/reader.go

```diff
@@ -0,0 +1,471 @@
+			return nil, err
+		}
+		if err := pl.Union(nextPl); err != nil {
+			return nil, err
+		}
+
+		iterErr = iter.Next()
+	}
+
+	if err := iterCloser.Close(); err != nil {
+		return nil, err
+	}
+
+	if err := fstCloser.Close(); err != nil {
+		return nil, err
+	}
+
+	return pl, nil
+}
+
+func (r *reader) MatchAll() (postings.MutableList, error) {
+	return nil, errNotImplemented
+}
+
+func (r *reader) Docs(pl postings.List) (doc.Iterator, error) {
+	return nil, errNotImplemented
+}
+
+func (r *reader) AllDocs() (doc.Iterator, error) {
+	return nil, errNotImplemented
+}
+
+func (r *reader) retrievePostingsList(postingsOffset uint64) (postings.List, error) {
+	postingsBytes, err := r.retrieveBytes(r.opts.PostingsData, postingsOffset)
+	if err != nil {
+		return nil, fmt.Errorf("unable to retrieve postings data: %v", err)
+	}
+
+	return pilosa.Unmarshal(postingsBytes, roaring.NewPostingsList)
+}
+
+func (r *reader) allKeys(fst *vellum.FST) ([][]byte, error) {
+	num := fst.Len()
+	keys := make([][]byte, 0, num)
+
+	iter, iterErr := fst.Iterator(minByteKey, maxByteKey)
+	safeCloser := x.NewSafeCloser(iter)
+	defer safeCloser.Close()
+
+	for {
+		if iterErr == vellum.ErrIteratorDone {
+			break
+		}
+		if iterErr != nil {
+			return nil, iterErr
+		}
+		key, _ := iter.Current()
+		keys = append(keys, r.copyBytes(key))
+		iterErr = iter.Next()
+	}
+
+	if err := safeCloser.Close(); err != nil {
+		return nil, err
+	}
+	return keys, nil
+}
+
+func (r *reader) retrieveTermsFST(field []byte) (*vellum.FST, error) {
+	termsFSTOffset, exists, err := r.fieldsFST.Get(field)
+	if err != nil {
+		return nil, err
+	}
+
+	if !exists {
+		return nil, fmt.Errorf("no terms known for field: %v", string(field))
+	}
+
+	termsFSTBytes, err := r.retrieveBytes(r.opts.FSTTermsData, termsFSTOffset)
+	if err != nil {
+		return nil, fmt.Errorf("error while decoding terms fst: %v", err)
+	}
+
+	return vellum.Load(termsFSTBytes)
+}
+
+// retrieveBytes assumes the base []byte slice is a collection of (payload, size, magicNumber) triples,
+// where size/magicNumber are strictly uint64 (i.e. 8 bytes). It assumes the 8 bytes preceding the offset
+// are the magicNumber, the 8 bytes before that are the size, and the `size` bytes before that are the
+// payload. It retrieves the payload while doing bounds checks to ensure no segfaults.
+func (r *reader) retrieveBytes(base []byte, offset uint64) ([]byte, error) {
+	const sizeofUint64 = 8
+	var (
+		magicNumberEnd   = int64(offset) // to prevent underflows
+		magicNumberStart = offset - sizeofUint64
+	)
+	if magicNumberEnd > int64(len(base)) || magicNumberStart < 0 {
+		return nil, fmt.Errorf("base bytes too small, length: %d, base-offset: %d", len(base), magicNumberEnd)
+	}
+	magicNumberBytes := base[magicNumberStart:magicNumberEnd]
+	d := encoding.NewDecoder(magicNumberBytes)
+	n, err := d.Uint64()
+	if err != nil {
+		return nil, fmt.Errorf("error while decoding magicNumber: %v", err)
+	}
+	if n != uint64(magicNumber) {
```

Comment:
Why include a magic number in each tuple?

Reply:
talked offline: used to delineate object boundaries.